<a href="https://colab.research.google.com/github/Ted-star7/Liquor-store-data-segregation/blob/main/Liquor_store_data_segregation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Uzapoint Liquor Data Classification & Cleansing Framework
## Data Integrity, Taxonomy Standardization & POS Optimization

**Author:** Teddy Kibuthu  
**Organization:** Uzapoint  
**Objective:** Standardize liquor product taxonomy, improve data integrity, and design a clean classification structure for POS optimization.

---

## 1. Executive Summary

This project focuses on cleaning, standardizing, and restructuring liquor product data to improve classification accuracy, onboarding simplicity, and reporting reliability within the Uzapoint POS system.

The dataset currently contains product-level classification fields including:

- Category
- Subcategory
- Product ID
- Product Label
- Product Image

However, inconsistencies in naming, classification structure, duplication, and missing image references may reduce data quality and operational efficiency.

This notebook establishes:

1. A structured cleaning pipeline
2. A standardized liquor taxonomy framework
3. A segregation model for anomaly detection
4. Business intelligence insights for POS enhancement
5. A final clean dataset ready for Power BI and POS integration

---

## 2. Business Objectives

The primary goals of this analysis are:

- Improve product classification consistency
- Detect and segregate anomalies
- Simplify liquor store onboarding structure
- Enhance POS search and filtering performance
- Prepare a clean, BI-ready dataset
- Enable future intelligent automation within Uzapoint

---

## 3. Project Roadmap

This notebook will proceed in structured stages:

1. Data Loading
2. Initial Data Audit
3. Data Cleaning & Standardization
4. Segregation of Anomalies
5. Image Integrity Analysis
6. Taxonomy Restructuring
7. Business Intelligence Insights
8. POS Integration Opportunities
9. Final Clean Dataset Export

## 4. Data Loading

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 4. Data Loading

This section loads the liquor product dataset from Google Drive into the Colab environment.

The dataset contains the following core fields:

- Category
- Subcategory
- Product ID
- Product Label
- Product Image (URL)

The objective of this stage is to:

- Load the dataset
- Validate its structure
- Confirm row count
- Preview initial records

In [3]:
import pandas as pd

#define file path
file_path = "/content/drive/My Drive/Dataset/Liquor Store Products.csv"

#load the csv
df = pd.read_csv(file_path)

#display the first 5 rows
print("Successfully loaded data. Here are the first 5 rows:")
print(df.head())

Successfully loaded data. Here are the first 5 rows:
  Category Subcategory  Product ID           Product Label  \
0    BEERS        CANS       10327  Tusker Lager Can 500ml   
1    BEERS        CANS       10328   Tusker Malt Can 500ml   
2    BEERS        CANS       10329   Tusker Lite Can 500ml   
3    BEERS        CANS       10330  Tusker Cider Can 500ml   
4    BEERS        CANS       10331     Guinness Can 500 Ml   

                                       Product Image  
0  https://uzapointerp.uzahost.com/uploads/produc...  
1  https://uzapointerp.uzahost.com/uploads/produc...  
2  https://uzapointerp.uzahost.com/uploads/produc...  
3  https://uzapointerp.uzahost.com/uploads/produc...  
4  https://uzapointerp.uzahost.com/uploads/produc...  


In [4]:
#Checking the Data Shape
print("Dataset shape:", df.shape)

#Checking column names
print("\ncolumns")
print(df.columns)

#Checking data info
print("\nDataset info")
print(df.info())


Dataset shape: (47944, 5)

columns
Index(['Category', 'Subcategory', 'Product ID', 'Product Label',
       'Product Image'],
      dtype='object')

Dataset info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47944 entries, 0 to 47943
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Category       47944 non-null  object
 1   Subcategory    47944 non-null  object
 2   Product ID     47944 non-null  int64 
 3   Product Label  47944 non-null  object
 4   Product Image  7977 non-null   object
dtypes: int64(1), object(4)
memory usage: 1.8+ MB
None


### 4.1 Data Structure Observations

The dataset contains:

- **Total Rows:** 47,944
- **Total Columns:** 5
- All core classification fields are complete (no nulls)
- Product Image column has significant missing values

#### Image Completeness Insight

Only 7,977 out of 47,944 products have image URLs.

This means approximately 83% of products lack image references.

This has implications for:

- POS user interface experience
- Product identification accuracy
- Duplicate detection using image similarity
- Onboarding simplicity

The next phase will focus on conducting a structured Data Audit.

## 5. Initial Data Audit

This section evaluates:

- Duplicate Product IDs
- Duplicate Product Labels
- Category consistency
- Subcategory consistency
- Image completeness ratio
- Whitespace and formatting inconsistencies

In [13]:
#check for duplicates in products Id
duplicates_id = df['Product ID'].duplicated().sum()

#check for duplicates in products label
duplicates_label = df['Product Label'].duplicated().sum()

# Check nulls for all columns
null_counts = df.isnull().sum()

#unique categorize
unique_categories = df['Category'].nunique()

#unique Subcategories
unique_subcategories = df['Subcategory'].nunique()

#Missing Images
missing_images = df['Product Image'].isnull().sum()
image_coverage = (1 - (missing_images / len(df))) * 100

print("Duplicate Product IDs:", duplicates_id)
print("Duplicate Product Labels:", duplicates_label)
print("Unique Categories:", unique_categories)
print("Unique Subcategories:", unique_subcategories)
print("Missing Images:", missing_images)
print("Image Coverage %:", image_coverage)
print("Null Values per Column:\n", null_counts)


Duplicate Product IDs: 1
Duplicate Product Labels: 15190
Unique Categories: 391
Unique Subcategories: 1045
Missing Images: 39967
Image Coverage %: 16.638161188052724
Null Values per Column:
 Category                0
Subcategory             0
Product ID              0
Product Label           0
Product Image       39967
Category_norm           0
Subcategory_norm        0
dtype: int64


In [6]:
# core metrics
total_rows = len(df)
duplicate_ids = df['Product ID'].duplicated().sum()
duplicate_labels = df['Product Label'].duplicated().sum()
missing_images = df['Product Image'].isnull().sum()

# Create data quality scorecard
data_quality_metrics = {
    'Metric': [
        'Completeness (Category)',
        'Completeness (Subcategory)',
        'Completeness (Product ID)',
        'Completeness (Product Label)',
        'Completeness (Product Image)',
        'Uniqueness (Product ID)',
        'Uniqueness (Product Label)'
    ],
    'Score (%)': [
        100,
        100,
        100,
        100,
        round((1 - missing_images/total_rows)*100, 2),
        round((1 - duplicate_ids/total_rows)*100, 2),
        round((1 - duplicate_labels/total_rows)*100, 2)
    ]
}

quality_df = pd.DataFrame(data_quality_metrics)

# Add status indicator
quality_df['Status'] = quality_df['Score (%)'].apply(
    lambda x: 'Excellent' if x >= 95
    else ' Moderate' if x >= 80
    else ' Critical'
)

print("Data Quality Scorecard")
quality_df

Data Quality Scorecard


Metric  Score (%)     Status
0       Completeness (Category)     100.00  Excellent
1    Completeness (Subcategory)     100.00  Excellent
2     Completeness (Product ID)     100.00  Excellent
3  Completeness (Product Label)     100.00  Excellent
4  Completeness (Product Image)      16.64   Critical
5       Uniqueness (Product ID)     100.00  Excellent
6    Uniqueness (Product Label)      68.32   Critical

## 5.1 Executive Data Quality Summary

### Overview

The dataset contains **47,944 liquor products** across 5 structural fields.  
While completeness across core columns is strong, significant classification and naming inconsistencies limit analytical and operational effectiveness.

Overall, the dataset demonstrates **high structural completeness but low classification control**.

---

### Key Findings

#### 1️ Product Label Duplication – 31.68%

15,190 product labels are duplicated.

This indicates:

- Identical products entered multiple times  
- Naming inconsistencies (format variations)  
- Missing distinguishing details within labels  

**Business Impact:**

- Split sales reporting  
- Inventory mismatches  
- Duplicate stock ordering  
- Customer confusion at POS  

---

#### 2️ Category Fragmentation – 391 Unique Categories

A liquor retail environment typically operates within a controlled set of top-level categories (commonly fewer than 15).

The presence of 391 categories suggests uncontrolled classification entry.

**Business Impact:**

- Overloaded POS dropdowns  
- Broken category-level reporting  
- Inconsistent performance analytics  
- Complex onboarding for new stores  

---

#### 3️ Subcategory Explosion – 1,045 Unique Subcategories

The large number of subcategories indicates potential structural misuse of classification layers.

Subcategories may be absorbing product-level variations rather than representing standardized grouping logic.

**Business Impact:**

- Fragmented analytics  
- Reduced search efficiency  
- Inconsistent grouping in reporting  
- Difficult aggregation of performance metrics  

---

#### 4️ Image Coverage – 16.64%

Only 7,977 out of 47,944 products contain image URLs.

This significantly limits:

- POS visual navigation  
- Customer-assisted product identification  
- Image-based duplicate detection  
- UI usability  

---

#### 5️ Product ID Integrity – Strong

Only 1 duplicate Product ID was detected.

This indicates a stable primary key structure and provides a reliable foundation for taxonomy reconstruction.

---

### Strategic Conclusion

The dataset is operationally usable but analytically fragmented.

The next phase will focus on:

- Category normalization  
- Subcategory standardization  
- Controlled taxonomy reconstruction  
- Structured segregation of anomalies  

## 6. Category Normalization & Structural Profiling

This section evaluates whether the high number of categories and subcategories
is driven by formatting inconsistencies (case sensitivity, whitespace) or by
true structural misclassification.

The objective is to:

- Normalize category and subcategory text
- Measure category collapse after normalization
- Quantify structural vs cosmetic classification issues

In [8]:
#create a normalized version
df ['Category_norm'] = (
    df['Category']
    .str.lower()
    .str.strip()
)
df ['Subcategory_norm'] =(
    df['Subcategory']
    .str.lower()
    .str.strip()
)
#compare original vs normalized
original_categories = df['Category'].nunique()
normalized_categories = df['Category_norm'].nunique()

original_subcategories = df['Subcategory'].nunique()
normalized_subcategories = df['Subcategory_norm'].nunique()

print("Original Categories:", original_categories)
print("Normalized Categories:", normalized_categories)
print("Category Reduction:", original_categories - normalized_categories)

print("Original Subcategories:", original_subcategories)
print("Normalized Subcategories:", normalized_subcategories)
print("Category Reduction:", original_subcategories - normalized_categories)

Original Categories: 391
Normalized Categories: 330
Category Reduction: 61
Original Subcategories: 1045
Normalized Subcategories: 921
Category Reduction: 715


In [11]:
#categories that fragment most after normalisation
category_fragmentation =(
    df.groupby('Category_norm')
    ['Category']
    .nunique()
    .sort_values(ascending=False)
)
category_fragmentation.head(15)

Category_norm
spirits        4
beers          3
food           3
wine           3
whiskey        3
soft drink     3
soft drinks    3
whisky         3
cigarettes     3
gin            3
juice          2
chocolates     2
cereals        2
water          2
sodas          2
Name: Category, dtype: int64

## 6.2 Category & Subcategory Normalization Findings

### 6.2.1 Category Standardization Impact

- **Original Categories:** 391  
- **Normalized Categories:** 330  
- **Total Reduction:** 61  

This indicates that duplicate and inconsistent category labels were successfully consolidated into standardized forms.

---

### 6.2.2 Subcategory Standardization Impact

- **Original Subcategories:** 1,045  
- **Normalized Subcategories:** 921  
- **Total Reduction:** 124  

This confirms that inconsistent subcategory naming was reduced, improving aggregation accuracy.

---

### 6.2.3 Most Frequent Normalized Categories

| Category        | Count |
|----------------|--------|
| spirits        | 4      |
| beers          | 3      |
| food           | 3      |
| wine           | 3      |
| whiskey        | 3      |
| soft drink     | 3      |
| soft drinks    | 3      |
| whisky         | 3      |
| cigarettes     | 3      |
| gin            | 3      |
| juice          | 2      |
| chocolates     | 2      |
| cereals        | 2      |
| water          | 2      |
| sodas          | 2      |

---

### 6.2.4 Key Observations

- Minor naming inconsistencies still exist (e.g., *whiskey vs whisky*, *soft drink vs soft drinks*).
- Further harmonization can improve analytical precision.
- Normalization enhances dashboard accuracy and business insight reliability.